Firestore/core/src/model/mutable_document.cc

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/model/mutable_document.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+/* static */
+MutableDocument MutableDocument::InvalidDocument(
+    const firebase::firestore::model::DocumentKey& document_key) {
+  return MutableDocument{document_key, DocumentType::kInvalid,
+                         SnapshotVersion::None(), ObjectValue{},
+                         DocumentState::kSynced};
+}
+
+/* static */
+MutableDocument MutableDocument::FoundDocument(
+    const firebase::firestore::model::DocumentKey& document_key,
+    const firebase::firestore::model::SnapshotVersion& version,
+    firebase::firestore::model::ObjectValue value) {
+  return std::move(InvalidDocument(document_key)
```

Comment:
You can remove the `std::move` on return statements, applied elsewhere too. See https://en.cppreference.com/w/cpp/language/copy_elision

Reply:
This doesn't seem to work. The compiler complains here since there is no copy constructor. I suspect that by default it tries to use the copy constructor and it wants to be free to do so, even if it end up using copy elision later on.

Reply:
Interesting, I learnt something!